From aa40854b3da17f48960540d74902cff0b27c4870 Mon Sep 17 00:00:00 2001
From: vmakarov <vmakarov@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Fri, 12 Apr 2013 17:13:45 +0000
Subject: [PATCH] 2013-04-12  Vladimir Makarov  <vmakarov@redhat.com>

	PR target/56903
	* config/i386/i386.c (ix86_hard_regno_mode_ok): Add
	lra_in_progress for return.

2013-04-12  Vladimir Makarov  <vmakarov@redhat.com>

	PR target/56903
	* gcc.target/i386/pr56903.c: New test.



git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@197928 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog                           |    6 ++++++
 gcc/config/i386/i386.c                  |    5 +++++
 gcc/testsuite/ChangeLog                 |    5 +++++
 gcc/testsuite/gcc.target/i386/pr56903.c |   18 ++++++++++++++++++
 4 files changed, 34 insertions(+), 0 deletions(-)
 create mode 100644 gcc/testsuite/gcc.target/i386/pr56903.c

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index 111c0a0..33ce936 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,9 @@
+2013-04-12  Vladimir Makarov  <vmakarov@redhat.com>
+
+	PR target/56903
+	* config/i386/i386.c (ix86_hard_regno_mode_ok): Add
+	lra_in_progress for return.
+
 2013-04-12  Jakub Jelinek  <jakub@redhat.com>
 
 	PR tree-optimization/56918
diff --git a/gcc/config/i386/i386.c b/gcc/config/i386/i386.c
index 63bf275..30d6816 100644
--- a/gcc/config/i386/i386.c
+++ b/gcc/config/i386/i386.c
@@ -33862,6 +33862,11 @@ ix86_hard_regno_mode_ok (int regno, enum machine_mode mode)
 	return true;
       if (!TARGET_PARTIAL_REG_STALL)
 	return true;
+      /* LRA checks if the hard register is OK for the given mode.
+        QImode values can live in non-QI regs, so we allow all
+        registers here.  */
+      if (lra_in_progress)
+       return true;
       return !can_create_pseudo_p ();
     }
   /* We handle both integer and floats in the general purpose registers.  */
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index 11d8ef0..25f29ab 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,8 @@
+2013-04-12  Vladimir Makarov  <vmakarov@redhat.com>
+
+	PR target/56903
+	* gcc.target/i386/pr56903.c: New test.
+
 2013-04-12  Jakub Jelinek  <jakub@redhat.com>
 
 	PR tree-optimization/56918
diff --git a/gcc/testsuite/gcc.target/i386/pr56903.c b/gcc/testsuite/gcc.target/i386/pr56903.c
new file mode 100644
index 0000000..9e6a1c3
--- /dev/null
+++ b/gcc/testsuite/gcc.target/i386/pr56903.c
@@ -0,0 +1,18 @@
+/* PR rtl-optimization/56903 */
+/* { dg-do compile } */
+/* { dg-options "-Os" } */
+/* { dg-additional-options "-march=pentium3" { target ia32 } } */
+
+int a, *b, c;
+struct S { int s : 1; } *fn1 (void);
+extern int fn3 (void), fn4 (int *);
+
+void
+fn2 (void)
+{
+  int e = fn3 ();
+  char f = c + fn1 ()->s * 4;
+  if (*b && f == e)
+    a = *b;
+  fn4 (b);
+}
-- 
1.7.1

